### PR TITLE
ReportVectorMigration changes

### PR DESCRIPTION
**Pull request by stitova-idm**
Implementing https://github.com/InstituteforDiseaseModeling/DtkTrunk/issues/4976

### CODE REVIEW

File: baseReportLib/report_params.rc [hunk 1]
```diff
@@ -95,3 +95,9 @@
...
 #define RISM_Include_Data_Threshold_Gametocytes_DESC_TEXT "If the column is included and the number of gametocytes related to this infection are greater than or equal to this value, then the row of data will be included."
+
+#define RVM_Must_Be_In_State_DESC_TEXT "A comma-separated list of vector states for which you want to record the migration. Note: Only STATE_MALE, STATE_ADULT, STATE_INFECTED, STATE_INFECTIOUS migrate."
+#define RVM_Must_Be_From_Node_DESC_TEXT "A comma-separated list of node IDs. A vector must be travelling FROM one of these nodes to be recorded into the report. Empty list means vectors traveling from any/all nodes will be recorded."
+#define RVM_Must_Be_To_Node_DESC_TEXT "A comma-separated list of node IDs. A vector must be travelling TO one of these nodes to be recorded into the report. Empty list means vectors traveling to any/all nodes will be recorded."
+#define RVM_Include_Genome_Data_DESC_TEXT "When set to 1, adds a Genome column for migrating vectors. Vectors with no custom genome will still have their sex alleles listed."
```
Comment (Bridenbecker) [inline baseReportLib/report_params.rc:102]: Maybe change to:

When set to **true (1)**, adds a Genome column for migrating vectors.  Vectors with no custom **alleles** will still have their sex alleles listed.

The "set to true (1)" is the pattern the doc team has used.

File: baseReportLib/report_params.rc [hunk 2]
```diff
@@ -95,3 +95,9 @@
 #define RISM_Include_Data_Threshold_Hepatocytes_DESC_TEXT "If the column is included and the number of hepatocytes related to this infection are greater than or equal to this value, then the row of data will be included."
 #define RISM_Include_Data_Threshold_IRBC_DESC_TEXT "If the column is included and the number of IRBCs related to this infection are greater than or equal to this value, then the row of data will be included."
 #define RISM_Include_Data_Threshold_Gametocytes_DESC_TEXT "If the column is included and the number of gametocytes related to this infection are greater than or equal to this value, then the row of data will be included."
+
+#define RVM_Must_Be_In_State_DESC_TEXT "A comma-separated list of vector states for which you want to record the migration. Note: Only STATE_MALE, STATE_ADULT, STATE_INFECTED, STATE_INFECTIOUS migrate."
```
Comment (Bridenbecker) [inline baseReportLib/report_params.rc:99]: Hmm.  I'm not sure we say "comma-separated" anywhere else.  I think we just say "list"

Comment (Bridenbecker) [reply baseReportLib/report_params.rc:99]: Probably don't need "Note:"

File: reporters/ReportVectorMigration.cpp [hunk 3]
```diff
@@ -93,10 +98,16 @@ GetReportInstantiator( Kernel::instantiator_function_t* pif )
...
+        : BaseTextReport(DEFAULT_NAME)
+        , m_StartDay( 0)
+        , m_EndDay(FLT_MAX)
         , m_IsValidDay( false )
+        , m_MustBeInState()
+        , m_MustBeFromNode()
+        , m_MustBeToNode()
+        , m_IncludeGenomeData( false )
+        , m_SpeciesList()
+        , m_FilenameSuffix()
```
Comment (Bridenbecker) [inline reporters/ReportVectorMigration.cpp:110]: Is it better that we copy all this filtering stuff here or someone use ReportFilter?  StartDay, EndDay, FilenameSuffix are all in ReportFilter.  I also wonder if we had a ReportFilterVector, could we standardize our report filtering.

File: reporters/ReportVectorMigration.cpp [hunk 4]
```diff
@@ -125,6 +145,42 @@ GetReportInstantiator( Kernel::instantiator_function_t* pif )
...
+                    case VectorStateEnum::STATE_LARVA:
+                    {
+                        throw InvalidInputDataException(__FILE__, __LINE__, __FUNCTION__,
+                            "'STATE_LARVA' is not a valid state for ReportVectorMigration.\n");
+                    }
+                    case VectorStateEnum::STATE_IMMATURE:
+                    {
+                        throw InvalidInputDataException(__FILE__, __LINE__, __FUNCTION__,
+                            "'STATE_IMMATURE' is not a valid state for ReportVectorMigration.\n");
+                    }
```
Comment (Bridenbecker) [inline reporters/ReportVectorMigration.cpp:166]: I think you could change this to something like:

                    case VectorStateEnum::STATE_EGG:
                    case VectorStateEnum::STATE_LARVA:
                    case VectorStateEnum::STATE_IMMATURE:
                    {
                        std::stringstream ss;
                        ss << "'" << VectorStateEnum::pairs::lookupKey( state ) << "'" is not a validate for ReportVectorMigration.";
                        throw InvalidInputDataException(__FILE__, __LINE__, __FUNCTION__, ss.msg().c_str );
                    }

File: reporters/ReportVectorMigration.cpp [hunk 5]
```diff
@@ -125,6 +145,42 @@ GetReportInstantiator( Kernel::instantiator_function_t* pif )
...
+                    case VectorStateEnum::STATE_MALE:
+                    case VectorStateEnum::STATE_ADULT:
+                    case VectorStateEnum::STATE_INFECTIOUS:
+                    case VectorStateEnum::STATE_INFECTED:
+                        break;
+                }
+            }
+            if (!m_FilenameSuffix.empty())
+            {
+                IdmString tmp_report_name = DEFAULT_NAME;
```
Comment (Bridenbecker) [inline reporters/ReportVectorMigration.cpp:176]: I would use m_ReportName and not DEFAULT_NAME

File: reporters/ReportVectorMigration.cpp [hunk 6]
```diff
@@ -125,6 +145,42 @@ GetReportInstantiator( Kernel::instantiator_function_t* pif )
...
+                std::vector<IdmString> parts = tmp_report_name.split('.');
+                release_assert(parts.size() == 2);
+                std::string output_fn = parts[0];
+                output_fn += "_" + m_FilenameSuffix + ".csv";
+                SetReportName(output_fn);
```
Comment (Bridenbecker) [inline reporters/ReportVectorMigration.cpp:181]: We should standardized this.

File: reporters/ReportVectorMigration.cpp [hunk 7]
```diff
@@ -166,33 +227,60 @@ GetReportInstantiator( Kernel::instantiator_function_t* pif )
             throw QueryInterfaceException(__FILE__, __LINE__, __FUNCTION__, "pvc", "IMigrate", "IVectorCohort");
         }
 
+        VectorStateEnum::Enum state = pvc->GetState();
+        if (!m_MustBeInState.empty())
+        {
+            if (!std::count(m_MustBeInState.begin(), m_MustBeInState.end(), state))
```
Comment (Bridenbecker) [inline reporters/ReportVectorMigration.cpp:233]: FYI. Normally, we use std::find() for something like this.  It would be faster if the list were very long.  In this, this is fine and not change is required.

File: reporters/ReportVectorMigration.cpp [hunk 8]
```diff
@@ -166,33 +227,60 @@ GetReportInstantiator( Kernel::instantiator_function_t* pif )
...
+                // from_node_id is not one we want in the report
+                return;
+            }
+        }
         int to_node_id = pSim->GetNodeExternalID( pim->GetMigrationDestination() ) ;
-        int mig_type = pim->GetMigrationType() ;
-        VectorStateEnum::Enum state = pvc->GetState();
+        if (!m_MustBeToNode.empty())
+        {
+            if (!(std::find(m_MustBeToNode.begin(), m_MustBeToNode.end(), to_node_id) != m_MustBeToNode.end()))
```
Comment (Bridenbecker) [inline reporters/ReportVectorMigration.cpp:253]: why did you switch from count to find?

I'm really wondering if there is a way to create a function that does this test.

File: utils/ConfigurationImpl.h [hunk 9]
```diff
@@ -167,6 +167,22 @@ namespace Kernel
...
+
+                for (int k = 0; k < enum_value_specs.size(); k++)
+                {
+                    qb["enum"][k] = json::String(enum_value_specs[k].first);
+                }
+
+                qb["description"] = json::String(description);
+                qb["default"] = json::Array();
+                return member;
+            }
```
Comment (Bridenbecker) [inline utils/ConfigurationImpl.h:184]: So I don't have to figure it out, why is this needed?

File: utils/version_info.h [hunk 10]
```diff
@@ -8,21 +8,21 @@ static const char* BUILD_DATE = __DATE__ " " __TIME__;
 #endif
 
 #ifndef SCCS_BRANCH
-#define    SCCS_BRANCH      "unknown-branch (unknown)"
+#define    SCCS_BRANCH      "ReportVectorMigration_changes_main (d8ed6a9d)"
```
Comment (Bridenbecker) [inline utils/version_info.h:11]: You need to remove this file

**Review by Bridenbecker — CHANGES_REQUESTED** [review]
I think there are somethings that could improve this code, but the one that that must be changed is not including version_info.h

File: reporters/ReportVectorMigration.cpp (hunk 6 above)
Comment (stitova-idm) [reply reporters/ReportVectorMigration.cpp:181]: I can use GetNewReportName for this, actually

Comment (stitova-idm) [reply reporters/ReportVectorMigration.cpp:181]: I got the code mostly from ReportFilter, but I don't think we can add the ReportFilter to ReportVectorMigration because it has so much human stuff in there? We could create a "baseReportFilter" class that's not human-oriented and then we can use that for start/stop, nodes, report name for the ReportFilter and ReportVectorMigration?

Comment (stitova-idm) [reply reporters/ReportVectorMigration.cpp:181]: or I guess add the report name functionality to the baseTextReport?

Comment (Bridenbecker) [reply reporters/ReportVectorMigration.cpp:181]: How about making task/ticket for you to look at this after vacation?  It should result in a refactor that has zero impact on user but reduces code reuse via copy-paste.

File: reporters/ReportVectorMigration.cpp (hunk 8 above)
Comment (stitova-idm) [reply reporters/ReportVectorMigration.cpp:253]: I will look at std::count vs find. i think this one is find because of your previous (on dtktrunk) comment. I'll standardize them, though.

File: utils/version_info.h (hunk 10 above)
Comment (stitova-idm) [reply utils/version_info.h:11]: lol. amateur hour.

File: reporters/ReportVectorMigration.cpp (hunk 6 above)
Comment (stitova-idm) [reply reporters/ReportVectorMigration.cpp:181]: https://github.com/InstituteforDiseaseModeling/EMOD/issues/29

File: reporters/ReportVectorMigration.cpp (hunk 8 above)
Comment (stitova-idm) [reply reporters/ReportVectorMigration.cpp:253]: moved everything to find without NOT

File: reporters/ReportVectorMigration.cpp (hunk 3 above)
Comment (stitova-idm) [reply reporters/ReportVectorMigration.cpp:110]: made it a TODO: https://github.com/InstituteforDiseaseModeling/EMOD/issues/29

File: reporters/ReportVectorMigration.cpp [hunk 11]
```diff
@@ -166,33 +220,61 @@ GetReportInstantiator( Kernel::instantiator_function_t* pif )
...
+        if (!m_MustBeToNode.empty())
+        {
+            if (std::find(m_MustBeToNode.begin(), m_MustBeToNode.end(), to_node_id) == m_MustBeToNode.end())
+            {
+                // to_node_id is not one we want in the report
+                return;
+            }
+        }
         std::string species = pvc->GetSpecies();
+        if (!m_SpeciesList.empty())
```
Comment (Bridenbecker) [inline reporters/ReportVectorMigration.cpp:254]: I'd change:

std::string species = pvc->GetSpecies();

to

const std::string& species = pvc->GetSpecies(); 

so we don't copy the string.

File: reporters/ReportVectorMigration.cpp [hunk 12]
```diff
@@ -166,33 +220,61 @@ GetReportInstantiator( Kernel::instantiator_function_t* pif )
...
+        {
+            if (std::find(m_SpeciesList.begin(), m_SpeciesList.end(), species) == m_SpeciesList.end())
+            {
+                // if not species we're interested in do not add to report
+                return;
+            }
+        }
```
Comment (Bridenbecker) [inline reporters/ReportVectorMigration.cpp:261]: Above you have the same code repeated four times.  What about replacing it with something like the following:

```
    template<typename T>
    bool Qualifies( const std::vector<T>& rList, const T& rValue )
    {
        if( rList.empty() )
        {
            return true;
        }
        else
        {
            return std::find( rList.begin(), rList.end(), rValue ) != rList.end();
        }
    }

    VectorStateEnum::Enum state        =  pvc->GetState();
    int                   from_node_id =  pSim->GetNodeExternalID( nodeSuid ) ;
    int                   to_node_id   =  pSim->GetNodeExternalID( pim->GetMigrationDestination() ) ;
    const std::string    species       =  pvc->GetSpecies(); 

    if( !Qualifies<ExternalNodeId_t     >( m_MustBeFromNode, from_node_id  ) ) return;
    if( !Qualifies<ExternalNodeId_t     >( m_MustBeToNode,   to_node_id    ) ) return;
    if( !Qualifies<VectorStateEnum::Enum>( m_MustBeInState,  state         ) ) return;
    if( !Qualifies<std::string          >( m_SpeciesList,    species       ) ) return;
```

**Review by Bridenbecker — CHANGES_REQUESTED** [review]
Still need to remove version_info.h from the project.  I provided some code that you might use.

File: utils/version_info.h [hunk 13]
```diff
@@ -8,21 +8,21 @@ static const char* BUILD_DATE = __DATE__ " " __TIME__;
 #endif
 
 #ifndef SCCS_BRANCH
-#define    SCCS_BRANCH      "ReportVectorMigration_changes_main (d8ed6a9d)"
```
[the remark this answers is not in the document]
Comment (stitova-idm) [reply utils/version_info.h]: it is no loger part of this pr (look at files changed)

**Review by Bridenbecker — APPROVED** [review]
LGTM assuming tests pass
